thrimbletrimmer/scripts/common.js

```diff
@@ -444,22 +418,192 @@ function dateTimeFromVideoPlayerTime(videoPlayerTime) {
+			return segment.start + dateTime.diff(segmentStart).as("seconds");
+		}
+	}
+	return null;
+}
+
+function videoHumanTimeFromDateTime(dateTime) {
+	const videoPlayerTime = videoPlayerTimeFromDateTime(dateTime);
+	if (videoPlayerTime === null) {
+		return null;
+	}
+	return videoHumanTimeFromVideoPlayerTime(videoPlayerTime);
+}
+
+function assembleVideoPlaylistURL(basePlaylistURL) {
+	let playlistURL = basePlaylistURL;
+
+	const queryStringParts = startAndEndTimeQueryStringParts();
+	if (queryStringParts) {
+		playlistURL += "?" + queryStringParts.join("&");
+	}
+	return playlistURL;
+}
+
+function startAndEndTimeQueryStringParts() {
+	const startTime = getStartTime();
+	const endTime = getEndTime();
+
+	let queryStringParts = [];
+	if (startTime) {
+		queryStringParts.push(`start=${wubloaderTimeFromDateTime(startTime)}`);
+	}
+	if (endTime) {
+		queryStringParts.push(`end=${wubloaderTimeFromDateTime(endTime)}`);
+	}
+	return queryStringParts;
+}
+
+function getSegmentList() {
+	return globalPlayer.latencyController.levelDetails.fragments;
+}
+
+function hasSegmentList() {
+	if (
+		globalPlayer &&
+		globalPlayer.latencyController &&
+		globalPlayer.latencyController.levelDetails &&
+		globalPlayer.latencyController.levelDetails.fragments
+	) {
+		return true;
+	}
+	return false;
+}
+
 function downloadFrame() {
 	const videoElement = document.getElementById("video");
 	const dateTime = dateTimeFromVideoPlayerTime(videoElement.currentTime);
-	const url = `/frame/${globalStreamName}/source.png?timestamp=${wubloaderTimeFromDateTime(dateTime)}`;
+	const url = `/frame/${globalStreamName}/source.png?timestamp=${wubloaderTimeFromDateTime(
+		dateTime
+	)}`;
 	// Avoid : as it causes problems on Windows
 	const filename = `${dateTime.toFormat("yyyy-LL-dd'T'HH-mm-ss.SSS")}.png`;
 	triggerDownload(url, filename);
 }
 
 function triggerDownload(url, filename) {
 	// URL must be same-origin.
-	const link = document.createElement('a');
-	link.setAttribute('download', filename);
+	const link = document.createElement("a");
+	link.setAttribute("download", filename);
 	link.href = url;
-	link.setAttribute('target', '_blank');
+	link.setAttribute("target", "_blank");
 	document.body.appendChild(link);
 	link.click();
 	document.body.removeChild(link);
 }
+
+async function getChatLog(startWubloaderTime, endWubloaderTime) {
+	globalChatData = [];
+	const url = `/${globalStreamName}/chat.json?start=${startWubloaderTime}&end=${endWubloaderTime}`;
+	const response = await fetch(url);
+	if (!response.ok) {
+		return;
+	}
+	const chatLogData = await response.json();
+	for (const chatLine of chatLogData) {
+		if (
+			chatLine.command !== "PRIVMSG" &&
+			chatLine.command !== "CLEARMSG" &&
+			chatLine.command !== "CLEARCHAT"
+		) {
+			continue;
+		}
+		const when = DateTime.fromSeconds(chatLine.time);
+		// Here, we just push each line successively into the list. This assumes data is provided to us in chronological order.
```

Comment:
Some messages may be out of order as we aren't certain of their correct order, we only have a time range they may have occurred in (and they'll be sorted by start-of-time-range order).
However, PRIVMSG, CLEARMSG and CLEARCHAT will always have an exact timestamp and so always be in order.

Reply:
In this case, I care specifically that `when` is in sequential order. If that's what matches with the unknown-here's-the-start-of-the-range timestamp (for other message types if I support them in the future), that's great.